File: frontend/lib/share.ts
/**
 * API client for the social sharing feature.
 *
 * All requests go through the Next.js API proxy at /api/v1/* which
 * handles service-to-service auth to the backend Cloud Run service.
 *
 * Follows the same patterns as api.ts — throws on non-2xx responses
 * with the error detail from the response body.
 */

const API_BASE = "/api/v1";

// ---------------------------------------------------------------------------
// Interfaces
// ---------------------------------------------------------------------------

export interface ParticipantSummary {
  role: string;
  name: string;
  agent_type: string;
  summary: string;
}

export interface SharePayload {
  share_slug: string;
  session_id: string;
  scenario_name: string;
  scenario_description: string;
  deal_status: "Agreed" | "Blocked" | "Failed";
  outcome_text: string;
  final_offer: number;
  turns_completed: number;
  warning_count: number;
  participant_summaries: ParticipantSummary[];
  elapsed_time_ms: number;
  share_image_url: string;
  created_at: string;
}

export interface SocialPostText {
  twitter: string;
  linkedin: string;
  facebook: string;
}

export interface CreateShareResponse {
  share_slug: string;
  share_url: string;
  social_post_text: SocialPostText;
  share_image_url: string;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

async function extractErrorDetail(res: Response): Promise<string> {
  try {
    const body = await res.json();
    return body.detail ?? body.message ?? JSON.stringify(body);
  } catch {
    return res.statusText || `HTTP ${res.status}`;
  }
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export async function createShare(
  sessionId: string,
  email: string,
): Promise<CreateShareResponse> {
  const res = await fetch(`${API_BASE}/share`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ session_id: sessionId, email }),
  });

  if (!res.ok) {
    const detail = await extractErrorDetail(res);
    throw new Error(detail);
  }

  return res.json();
}

export async function getShare(slug: string): Promise<SharePayload> {
  const res = await fetch(`${API_BASE}/share/${slug}`);

  if (!res.ok) {
    const detail = await extractErrorDetail(res);
    throw new Error(detail);
  }

  return res.json();
}

File: backend/app/models/share.py
"""Pydantic V2 models for the social sharing feature."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class ParticipantSummary(BaseModel):
    """Summary of a single negotiation participant."""

    role: str
    name: str
    agent_type: str
    summary: str


class EvaluationScores(BaseModel):
    """High-level evaluation dimension scores from the post-negotiation evaluator."""

    fairness: int = Field(ge=1, le=10)
    mutual_respect: int = Field(ge=1, le=10)
    value_creation: int = Field(ge=1, le=10)
    satisfaction: int = Field(ge=1, le=10)
    overall_score: int = Field(ge=1, le=10)


class PublicMessage(BaseModel):
    """A single public message from the negotiation conversation."""

    agent_name: str
    role: str
    agent_type: str
    message: str
    turn_number: int


class SharePayload(BaseModel):
    """Full persisted share document for a completed negotiation.

    Keyed by share_slug in the shared_negotiations collection/table.
    Contains only public-facing summary data — no raw history,
    hidden context, custom prompts, or model overrides.
    """

    share_slug: str = Field(min_length=8, max_length=8)
    session_id: str
    scenario_id: str = Field(default="")
    scenario_name: str
    scenario_description: str
    deal_status: Literal["Agreed", "Blocked", "Failed"]
    outcome_text: str
    final_offer: float = Field(ge=0)
    turns_completed: int = Field(ge=0)
    warning_count: int = Field(ge=0)
    participant_summaries: list[ParticipantSummary]
    evaluation_scores: EvaluationScores | None = Field(default=None)
    public_conversation: list[PublicMessage] = Field(default_factory=list)
    elapsed_time_ms: int = Field(ge=0)
    share_image_url: str
    created_at: datetime


class SocialPostText(BaseModel):
    """Platform-specific social media post text variants."""

    twitter: str = Field(max_length=280)
    linkedin: str = Field(max_length=3000)
    facebook: str = Field(max_length=3000)


class CreateShareRequest(BaseModel):
    """Request body for POST /api/v1/share."""

    session_id: str = Field(min_length=1)
    email: str = Field(min_length=1)


class CreateShareResponse(BaseModel):
    """Response body for POST /api/v1/share."""

    share_slug: str
    share_url: str
    social_post_text: SocialPostText
    share_image_url: str
